File: example implementations/java/convertor/src/voxel/XYZ.java
package voxel;

import utils.EL;
import static utils.NumberUtilities.encroach;

import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;

public class XYZ {
    public static final XYZ MIN_VALUE = new XYZ(Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MIN_VALUE);
    public static final XYZ MAX_VALUE = new XYZ(Integer.MAX_VALUE, Integer.MAX_VALUE, Integer.MAX_VALUE);
    public static final XYZ ZERO = new XYZ();
    public static final XYZ ONE = new XYZ(1);
    public int x;
    public int y;
    public int z;

    public XYZ(int i) {
        this(i, i, i);
    }

    public XYZ() {
        this(0, 0, 0);
    }

    public XYZ(byte x, byte y, byte z) {
        this(x & 0xff, y & 0xff, z & 0xff);
    }

    public XYZ(int x, int y, int z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    /**
     * Get x, y, or z respectively by their index.
     * Overflow is supported and will evaluate to modulo 3.
     * For example: index 3 evaluates to 0 and as such will return x.
     */
    public int get(int index) {
        int moddedIndex = index % 3;
        return switch (moddedIndex) {
            case 0 -> x;
            case 1 -> y;
            default -> z;
        };
    }

    /**
     * Set a value to x, y, or z respectively by their index.
     * Overflow is supported and will evaluate to modulo 3.
     * For example: index 3 evaluates to 0 and as such will set x.
     */
    public void set(int index, int value) {
        int moddedIndex = index % 3;
        switch (moddedIndex) {
            case 0 -> x = value;
            case 1 -> y = value;
            default -> z = value;
        }
    }

    /**
     * Both bounds are inclusive.
     */
    public boolean withinBounds(XYZ bound0, XYZ bound1) {
        XYZ lower = bound0.min(bound1);
        XYZ upper = bound0.max(bound1);
        return !hasLesserThan(lower) && !hasGreaterThan(upper);
    }

    public int sum() {
        return x + y + z;
    }

    public XYZ abs() {
        return transform(Math::abs);
    }

    public XYZ add(XYZ toAdd) {
        return combine(toAdd, Math::addExact);
    }

    public XYZ subtract(XYZ toSub) {
        return combine(toSub,Math::subtractExact);
    }

    public XYZ min(XYZ toMin) {
        return combine(toMin, Math::min);
    }

    public XYZ max(XYZ toMax) {
        return combine(toMax, Math::max);
    }

    public XYZ mult(XYZ toMult) {
        return combine(toMult, Math::multiplyExact);
    }

    /**
     * divide by, rounded down
     */
    public XYZ div(int divideBy) {
        return div(new XYZ(divideBy, divideBy, divideBy));
    }

    /**
     * divide by, rounded down
     */
    public XYZ div(XYZ divideBy) {
        return combine(divideBy, Math::divideExact);
    }

    public XYZ difference(XYZ toDiff) {
        return subtract(toDiff).abs();
    }

    public XYZ greaterThan(XYZ lesser) {
        return combine(lesser, (g, l) -> g > l ? 1 : 0);
    }

    public XYZ lesserThan(XYZ lesser) {
        return combine(lesser, (g, l) -> g < l ? 1 : 0);
    }

    public boolean has(Predicate<Integer> predicate) {
        return predicate.test(x) || predicate.test(y) || predicate.test(z);
    }

    public boolean hasLesserThan(XYZ greater) {
        return lesserThan(greater).sum() > 0;
    }

    public boolean hasGreaterThan(XYZ lesser) {
        return greaterThan(lesser).sum() > 0;
    }
    public XYZ combine(XYZ xyz, BiFunction<Integer, Integer, Integer> combiner) {
        return new XYZ(combiner.apply(x, xyz.x), combiner.apply(y, xyz.y), combiner.apply(z, xyz.z));
    }

    public XYZ transform(Function<Integer, Integer> transformer) {
        return new XYZ(transformer.apply(x), transformer.apply(y), transformer.apply(z));
    }
    // from inclusive, to exclusive
    public void fromTo(XYZ to, Consumer<XYZ> consumer) {
        for (int xx = x; xx != to.x; xx = encroach(xx, to.x))
            for (int yy = y; yy != to.y; yy = encroach(yy, to.y))
                for (int zz = z; zz != to.z; zz = encroach(zz, to.z))
                    consumer.accept(new XYZ(xx, yy, zz));
    }

    // from inclusive, to exclusive
    public EL<XYZ> fromToList(XYZ to, XYZ step) {
        EL<XYZ> res = new EL<>();
        for (int xx = x; xx != to.x; xx = encroach(xx, to.x, step.x))
            for (int yy = y; yy != to.y; yy = encroach(yy, to.y, step.y))
                for (int zz = z; zz != to.z; zz = encroach(zz, to.z, step.z)) {
                    res.add(new XYZ(xx, yy, zz));
                }
        return res;
    }

    @Override
    public boolean equals(Object obj) {
        return obj.getClass() == XYZ.class && ((XYZ) obj).x == x && ((XYZ) obj).y == y && ((XYZ) obj).z == z;
    }

    @Override
    public String toString() {
        return "voxel.XYZ{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                '}';
    }

    @Override
    public XYZ clone() {
        return new XYZ(x, y, z);
    }
}

File: example implementations/csharp/cvox-convertor/voxel/XYZ.cs
using static cvox_convertor.utils.NumberUtilities;

namespace cvox_convertor.voxel
{
    public class XYZ
    {
        public static readonly XYZ MIN_VALUE = new(int.MinValue, int.MinValue, int.MinValue);
        public static readonly XYZ MAX_VALUE = new(int.MaxValue, int.MaxValue, int.MaxValue);
        public static readonly XYZ ZERO = new();
        public int X;
        public int Y;
        public int Z;

        public XYZ(int i) => (X, Y, Z) = (i, i, i);

        public XYZ() => (X, Y, Z) = (0, 0, 0);

        public XYZ(byte x, byte y, byte z) => (X, Y, Z) = (x & 0xff, y & 0xff, z & 0xff);

        public XYZ(int x, int y, int z) => (X, Y, Z) = (x, y, z);

        public static bool operator ==(XYZ xyz0, XYZ xyz1)
        {
            return xyz0.X == xyz1.X && xyz0.Y == xyz1.Y && xyz0.Z == xyz1.Z;
        }

        public static bool operator !=(XYZ xyz0, XYZ xyz1)
        {
            return !(xyz0 == xyz1);
        }

        public static XYZ operator +(XYZ xyz0, XYZ xyz1)
        {
            return xyz0.Combine(xyz1, (a, b) => a + b);
        }

        public static XYZ operator -(XYZ xyz0, XYZ xyz1)
        {
            return xyz0.Combine(xyz1, (a, b) => a - b);
        }

        public static XYZ operator -(XYZ xyz0, int subtraction)
        {
            return xyz0.Transform(a => a - subtraction);
        }

        public static XYZ operator *(XYZ xyz0, XYZ xyz1)
        {
            return xyz0.Combine(xyz1, (a, b) => a * b);
        }

        public static XYZ operator /(XYZ xyz0, XYZ xyz1)
        {
            return xyz0.Combine(xyz1, (a, b) => a / b);
        }

        public static XYZ operator /(XYZ xyz, int divideBy)
        {
            return xyz.Transform(a => a / divideBy);
        }

        public static XYZ operator %(XYZ xyz0, XYZ xyz1)
        {
            return xyz0.Combine(xyz1, (a, b) => a % b);
        }

        public static XYZ operator %(XYZ xyz, int modBy)
        {
            return xyz.Transform(a => a % modBy);
        }

        /**
         * Get x, y, or z respectively by their index.
         * Overflow is supported and will evaluate to modulo 3.
         * For example: index 3 evaluates to 0 and as such will return x.
         */
        public int Get(int index)
        {
            int moddedIndex = index % 3;
            switch (moddedIndex)
            {
                case 0:
                    return X;
                case 1:
                    return Y;
                default:
                    return Z;
            };
        }

        /**
         * Set a value to x, y, or z respectively by their index.
         * Overflow is supported and will evaluate to modulo 3.
         * For example: index 3 evaluates to 0 and as such will set x.
         */
        public void Set(int index, int value)
        {
            int moddedIndex = index % 3;
            switch (moddedIndex)
            {
                case 0:
                    X = value;
                    break;
                case 1:
                    Y = value;
                    break;
                default:
                    Z = value;
                    break;
            }
        }

        /**
         * Both bounds are inclusive.
         */
        public bool WithinBounds(XYZ bound0, XYZ bound1)
        {
            XYZ lower = bound0.Min(bound1);
            XYZ upper = bound0.Max(bound1);
            return !HasLesserThan(lower) && !HasGreaterThan(upper);
        }

        public int Sum()
        {
            return X + Y + Z;
        }

        public XYZ Abs()
        {
            return Transform(Math.Abs);
        }

        public XYZ Min(XYZ toMin)
        {
            return Combine(toMin, Math.Min);
        }

        public XYZ Max(XYZ toMax)
        {
            return Combine(toMax, Math.Max);
        }

        public XYZ Difference(XYZ toDiff)
        {
            return (this - toDiff).Abs();
        }

        public XYZ GreaterThan(XYZ lesser)
        {
            return Combine(lesser, (g, l) => g > l ? 1 : 0);
        }

        public XYZ LesserThan(XYZ lesser)
        {
            return Combine(lesser, (g, l) => g < l ? 1 : 0);
        }

        public bool Has(Predicate<int> predicate)
        {
            return predicate(X) || predicate(Y) || predicate(Z);
        }

        public bool HasLesserThan(XYZ greater)
        {
            return LesserThan(greater).Sum() > 0;
        }

        public bool HasGreaterThan(XYZ lesser)
        {
            return GreaterThan(lesser).Sum() > 0;
        }
        public XYZ Combine(XYZ xyz, Func<int, int, int> combiner)
        {
            return new XYZ(combiner(X, xyz.X), combiner(Y, xyz.Y), combiner(Z, xyz.Z));
        }

        public XYZ Transform(Func<int, int> transformer)
        {
            return new XYZ(transformer(X), transformer(Y), transformer(Z));
        }
        // from inclusive, to exclusive
        public void FromTo(XYZ to, Action<XYZ> consumer)
        {
            for (int xx = X; xx != to.X; xx = Encroach(xx, to.X))
                for (int yy = Y; yy != to.Y; yy = Encroach(yy, to.Y))
                    for (int zz = Z; zz != to.Z; zz = Encroach(zz, to.Z))
                        consumer(new XYZ(xx, yy, zz));
        }
    }
}
